pennylane/hf/matrices.py

```diff
@@ -0,0 +1,346 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for computing matrices.
+"""
+
+import autograd.numpy as anp
+from pennylane.hf.integrals import (
+    generate_attraction,
+    generate_kinetic,
+    generate_overlap,
+    generate_repulsion,
+)
+
+
+def molecular_density_matrix(n_electron, c):
+    r"""Compute the molecular density matrix.
+
+    The density matrix :math:`P` is computed from the molecular orbital coefficients :math:`C` as
+
+    .. math::
+
+        P_{\mu \nu} = \sum_{i=1}^{N} C_{\mu i} C_{\nu i},
+
+    where :math:`N = N_{electrons} / 2` is the number of occupied orbitals. Note that the total
+    density matrix is the sum of the :math:`\alpha` and :math:`\beta` density
+    matrices, :math:`P = P^{\alpha} + P^{\beta}`.
+
+    Args:
+        n_electron (integer): number of electrons
+        c (array[float]): molecular orbital coefficients
+
+    Returns:
+        array[float]: density matrix
+
+    **Example**
+
+    >>> c = np.array([[-0.54828771,  1.21848441], [-0.54828771, -1.21848441]])
+    >>> n_electron = 2
+    >>> density_matrix(n_electron, c)
+    array([[0.30061941, 0.30061941], [0.30061941, 0.30061941]])
+    """
+    p = anp.dot(c[:, : n_electron // 2], anp.conjugate(c[:, : n_electron // 2]).T)
+    return p
+
+
+def overlap_matrix(basis_functions):
+    r"""Return a function that computes the overlap matrix for a given set of basis functions.
+
+    Args:
+        basis_functions (list[BasisFunction]): basis functions
+
+    Returns:
+        function: function that computes the overlap matrix
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]], requires_grad = False)
+    >>> alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+    >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True),
+    >>> mol = Molecule(symbols, geometry, alpha=alpha)
+    >>> args = [alpha]
+    >>> overlap_matrix(mol.basis_set)(*args)
+    array([[1.0, 0.7965883009074122], [0.7965883009074122, 1.0]])
+    """
+
+    def overlap(*args):
+        r"""Construct the overlap matrix for a given set of basis functions.
+
+        Args:
+            args (array[float]): initial values of the differentiable parameters
+
+        Returns:
+            array[float]: the overlap matrix
+        """
+        n = len(basis_functions)
+        s = anp.eye(len(basis_functions))
```

Comment:
You could maybe call it overlap_matrix, I think it would be clearer for the user.